include/ginkgo/core/config/property_tree.hpp

```diff
@@ -0,0 +1,169 @@
+IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
+TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
+PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#ifndef GKO_PUBLIC_CORE_CONFIG_PROPERTY_TREE_HPP_
+#define GKO_PUBLIC_CORE_CONFIG_PROPERTY_TREE_HPP_
+
+
+#include <cassert>
+#include <exception>
+#include <map>
+#include <stdexcept>
+#include <string>
+#include <type_traits>
+#include <vector>
+
+
+#include <ginkgo/core/config/data.hpp>
+
+
+namespace gko {
+namespace config {
+
+
+/**
+ * pnode is to describe the property tree
+ */
+class pnode {
+public:
+    using key_type = std::string;
+    using data_type = data;
+
+    enum status_t { empty, array, object, list };
+
+    pnode() : status_(status_t::empty) {}
+
+    pnode(const data_type& d) : status_(status_t::object), data_(d) {}
+
+    pnode(const std::vector<pnode>& array)
+        : status_(status_t::array), array_(array)
+    {}
+
+    pnode(const std::map<key_type, pnode>& list)
+        : status_(status_t::list), list_(list)
+    {}
+
+    // Get the content of node
+    data_type& get_data()
+    {
+        assert(status_ == status_t::object);
+        return data_;
+    }
+
+    const data_type& get_data() const
+    {
+        assert(status_ == status_t::object);
+        return data_;
+    }
+
+    std::vector<pnode>& get_array()
+    {
+        assert(status_ == status_t::array || status_ == status_t::empty);
+        status_ = status_t::array;
+        return array_;
+    }
+
+    const std::vector<pnode>& get_array() const
+    {
+        assert(status_ == status_t::array);
+        return array_;
+    }
+
+    std::map<key_type, pnode>& get_list()
+    {
+        assert(status_ == status_t::list || status_ == status_t::empty);
+        status_ = status_t::list;
+        return list_;
+    }
+
+    const std::map<key_type, pnode>& get_list() const
+    {
+        assert(status_ == status_t::list);
+        return list_;
+    }
+
+    // Get the data of node's content
+    template <typename T>
+    T get_data() const
+    {
+        assert(status_ == status_t::object);
+        return gko::config::get<T>(data_);
+    }
+
+    pnode& at(const std::string& path)
+    {
+        assert(status_ == status_t::list);
+        return list_.at(path);
+    }
+
+    const pnode& at(const std::string& path) const
```

Comment:
I think this (and the other at functions) should allow returning an empty `pnode`, if it doesn't contain the object. This, together with either `bool empty()` or a bool conversion operator, can allow easier code:
```
if(auto child = pt.at("name"){
  parse(child);
}
```
instead of repeating `pt.contains(...)` and `pt.at(...)`.

Reply:
TBH. I've not commented on that yet, but I would prefer only const access.

Reply:
currently, I return `pnode&` from `at()` and `const pnode&` from `at() const`. (For less copy in `auto& temp = pnode.at()`)
Return the empty node in `at()` is weird because it should not be modified and the modification does not change the content of property tree.
If only make it available in `at() const`, we need to cast it to const first for const `at`. However, it also introduces different non-found behavior.